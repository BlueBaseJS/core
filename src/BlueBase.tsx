import { Analytics, Logger } from './api';
import {
	ComponentInputCollection,
	ComponentRegistry,
	ConfigRegistry,
	ConfigsCollection,
	HookCollection,
	HookRegistry,
	PluginRegistry,
	ThemeItemCollection,
	ThemeProvider,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import { ComponentRegistryWithUIInterfaces as IComponentRegistry } from './ui-interfaces';
import { MaybeBlueBaseModule } from './utils';
import { Plugin } from './models/Plugin';
import React from 'react';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentInputCollection,
	configs: ConfigsCollection,
	hooks: HookCollection,
	plugins: Array<MaybeBlueBaseModule<Plugin>>,
	themes: ThemeItemCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components: IComponentRegistry = new ComponentRegistry(this) as IComponentRegistry;
	public Configs: ConfigRegistry = new ConfigRegistry(this);
	public Hooks: HookRegistry = new HookRegistry(this);
	public Plugins: PluginRegistry = new PluginRegistry(this);
	public Themes: ThemeRegistry = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerCollection(systemHooks);
		await this.Hooks.registerCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<this.Components.SystemApp>
						{options && options.children}
					</this.Components.SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}