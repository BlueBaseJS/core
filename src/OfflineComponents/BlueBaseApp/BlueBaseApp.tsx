import { BlueBase, BlueBaseProgress, BootOptions } from '../../BlueBase';
import {
	BlueBaseAppError,
	BlueBaseAppErrorProps,
	BlueBaseAppLoading,
	BlueBaseAppLoadingProps,
} from '..';
import React, { useEffect, useState } from 'react';

import { BlueBaseProvider } from '../../Context';
import { IntlProvider } from '../../intl';
import { ThemeProvider } from '../../themes';
import { getComponent } from '../../getComponent';

const BlueBaseContent = getComponent('BlueBaseContent');

export interface BlueBaseAppProps extends Partial<BootOptions> {
	/** BlueBase Context */
	BB?: BlueBase;

	/** Component rendered when an error occurs */
	ErrorComponent: React.ComponentType<BlueBaseAppErrorProps>;

	/** Component rendered when an error occurs */
	LoadingComponent: React.ComponentType<BlueBaseAppLoadingProps>;

	/**
	 * If this prop is provided, this tree will be rendered instead of BlueBase's own view.
	 */
	children?: React.ReactNode;

	/**
	 * Used to locate this view in end-to-end tests.
	 */
	testID?: string;
}

/**
 * # ðŸš€ BlueBaseApp
 *
 * The main BlueBase app. This is the top level component in BlueBase. Takes care
 * of initialisation, and renders either children, or app with routing.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseApp BB={BB} />
 * ```
 */
export const BlueBaseApp = (props: BlueBaseAppProps) => {
	const {
		ErrorComponent = BlueBaseAppError,
		LoadingComponent = BlueBaseAppLoading,
		children,
	} = props;

	const [bootCount, setBootCount] = useState(0);
	const [booting, setBooting] = useState(true);
	const [bootTrigger, setBootTrigger] = useState(true); // Setting to true to start boot
	const [progress, setProgress] = useState<BlueBaseProgress>({});

	const [BB] = useState(props.BB || new BlueBase());

	BB.reboot = async () => {
		setBootTrigger(true);
	};

	useEffect(() => {
		if (bootTrigger) {
			setBootTrigger(false);
		} else {
			return;
		}

		(async () => {
			await BB.boot({
				...props,
				onProgress: (p: BlueBaseProgress) => {
					setProgress(p);
					if (p.error) {
						setBooting(false);
					}
				},
			});
			setBootCount(bootCount + 1);
			setBooting(false);
		})();
	}, [bootTrigger]);

	if (booting) {
		return <LoadingComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	if (progress.error) {
		return <ErrorComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	return (
		<BlueBaseProvider key={`boot-${bootCount}`} value={BB}>
			<ThemeProvider>
				<IntlProvider>
					<BlueBaseContent>{children}</BlueBaseContent>
				</IntlProvider>
			</ThemeProvider>
		</BlueBaseProvider>
	);
};

const defaultProps: Partial<BlueBaseAppProps> = {
	ErrorComponent: BlueBaseAppError,
	LoadingComponent: BlueBaseAppLoading,
};

BlueBaseApp.defaultProps = defaultProps;
