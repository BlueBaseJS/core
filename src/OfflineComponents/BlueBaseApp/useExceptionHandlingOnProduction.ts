import { BlueBase } from '../../BlueBase';
import { isProduction } from '../../utils';

// Number of errors to save
const CACHE_SIZE = 5;

// If a successive error take longer than this time, we don't consider it as recursive
const RECURSIVE_PERIOD = 300;

let errors: Array<{
	error: any;
	time: number;
}> = [];

/**
 * If the app is in production mode, automatically reboots BB on exception.
 * Shows an error message if the app is stuck in infinite crash.
 * @param BB
 */
export function useExceptionHandlingOnProduction(BB: BlueBase) {
	let development = BB.Configs.getValue('development');

	if (development === undefined) {
		development = !isProduction();
	}

	/**
	 * Adds an error to the errors array, with the timestamp
	 * @param e
	 */
	function addError(e: any) {
		const arr = [
			...errors,
			{
				error: e,
				time: new Date().getTime(),
			},
		];

		errors = arr.length > CACHE_SIZE ? arr.slice(1) : arr;
	}

	/**
	 * Gets periods of time in successive errors
	 */
	function isInfinite() {
		if (errors.length < CACHE_SIZE) {
			return false;
		}

		for (let i = 1; i < errors.length; i++) {
			if (errors[i].time - errors[i - 1].time > RECURSIVE_PERIOD) {
				return false;
			}
		}

		return true;
	}

	function onError(e: any) {
		addError(e);

		if (!development && !isInfinite()) {
			BB.reboot();
		}
	}

	return { onError };
}
