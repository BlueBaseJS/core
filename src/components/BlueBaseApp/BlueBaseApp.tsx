import { BlueBase, BootOptions } from '../../BlueBase';
import React from 'react';
import { Text } from 'react-native';

export interface BlueBaseAppProps extends Partial<BootOptions> {
	BB?: BlueBase,
	children?: React.ReactNode,
}

interface BlueBaseAppState {

	booted: boolean;
	BB: BlueBase

	Component: React.ComponentType<any>
}

/**
 * ðŸš€ BlueBaseApp
 *
 * This is the Main App used to render at the top level.
 *
 * TODO: Add better docs
 */
// NOTE FOR DEVELOPERS:
// Don't use BlueBase context or any data saved in context (e.g. components)
// here. This is because the context may not be initialized yet.
export class BlueBaseApp extends React.Component<BlueBaseAppProps, BlueBaseAppState> {

	constructor(props: BlueBaseAppProps) {
		super(props);

		this.state = {
			BB: props.BB || new BlueBase(),
			Component: () => <Text>Loading</Text>,
			booted: false,
		};
	}

	async componentDidMount() {

		const BB = this.state.BB;

		try {
			const Component = await BB.boot(this.props);
			this.setState({
				Component,
				booted: BB.booted,
			});
		} catch (error) {
			// tslint:disable-next-line:no-console
			console.error(error);
		}
	}

	render() {

		const Component = this.state.Component;

		return <Component children={this.props.children} />;
	}
}