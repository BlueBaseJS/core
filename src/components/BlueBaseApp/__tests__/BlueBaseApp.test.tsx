import { BlueBaseApp } from '../BlueBaseApp';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { BlueBase } from '../../../BlueBase';

beforeEach(() => {
	jest.resetModules();
});

const Bang = () => {
	throw Error('ğŸ’¥ Boom!');
};

const BangNull = () => {
	throw null;
};

describe('BlueBaseApp', () => {

	test(`should render BlueBaseApp`, (done: any) => {
		const rendered: any = TestRenderer.create(<BlueBaseApp />);

		// Will show loading
		expect(rendered.toJSON().children.join()).toBe('Loading');
		expect(rendered).toMatchSnapshot();

		// After loading
		setTimeout(() => {
			expect(rendered).toMatchSnapshot();
			done();
		});
	});

	test(`should render default app when boot doesn't return anything`, (done: any) => {

		const BB = new BlueBase();
		BB.Configs.setValue('development', true);
		BB.boot = (() => { Promise.resolve(null); }) as any;

		const rendered: any = TestRenderer.create(
			<BlueBaseApp BB={BB} />
		);

		// After loading
		setTimeout(() => {
			const tree = rendered.toJSON();
			expect(tree).toBe(null);
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render error state when boot throws an error`, (done: any) => {

		const BB = new BlueBase();
		BB.Configs.setValue('development', true);
		BB.boot = () => { throw Error('Boot Error!'); };

		const rendered: any = TestRenderer.create(
			<BlueBaseApp BB={BB} />
		);

		// After loading
		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children.join()).toBe('Boot Error!');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render error state when a child throws an error`, (done: any) => {
		const rendered: any = TestRenderer.create(<BlueBaseApp><Bang /></BlueBaseApp>);

		// After loading
		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children.join()).toBe('ğŸ’¥ Boom!');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render error state with custom message when a child throws a null error`, (done: any) => {
		const rendered: any = TestRenderer.create(<BlueBaseApp><BangNull /></BlueBaseApp>);

		// After loading
		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children.join()).toBe('An unknown error occured.');
			expect(tree).toMatchSnapshot();
			done();
		});
	});

	test(`should render error state with custom message when in production mode`, (done: any) => {

		const rendered: any = TestRenderer.create(
			<BlueBaseApp configs={{ development: false }}>
				<Bang />
			</BlueBaseApp>
		);

		// After loading
		setTimeout(() => {
			const tree = rendered.toJSON();
			expect((tree as any).children.join()).toBe('An unknown error occured.');
			expect(tree).toMatchSnapshot();
			done();
		});
	});
});
