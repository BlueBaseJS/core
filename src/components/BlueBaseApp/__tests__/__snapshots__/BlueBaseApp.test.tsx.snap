// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlueBaseApp Snapshot BlueBaseApp component 1`] = `
<Text>
  Loading
</Text>
`;

exports[`BlueBaseApp Snapshot BlueBaseApp component after complete rendering 1`] = `
<View>
  <View>
    <Text
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      ðŸš€ BlueBase System Content!
    </Text>
  </View>
</View>
`;
